pkg/logs/build_logs.go

```diff
@@ -122,24 +122,20 @@ func (c *BuildLogsClient) watchReadyContainers(ctx context.Context, readyContain
 
 			switch r.Type {
 			case watch.Added, watch.Modified:
-
-				for _, c := range pod.Status.InitContainerStatuses {
-					if c.State.Waiting == nil {
-						readyContainers <- readyContainer{
+				containers := map[string]readyContainer{}
+				for _, c := range append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...) {
+					if c.State.Waiting == nil && buildapi.IsBuildStep(c.Name) {
+						containers[c.Name] = readyContainer{
 							podName:       pod.Name,
 							containerName: c.Name,
 							namespace:     pod.Namespace,
 						}
 					}
 				}
 
-				for _, c := range pod.Status.ContainerStatuses {
-					if c.State.Waiting == nil {
-						readyContainers <- readyContainer{
-							podName:       pod.Name,
-							containerName: c.Name,
-							namespace:     pod.Namespace,
-						}
+				for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+					if readyContainer, found := containers[container.Name]; found {
+						readyContainers <- readyContainer
 					}
 				}
```

Comment:
Is there an edge case where the logs are not streamed in order if a container becomes ready faster than the others?

E.g.
1. A new build pod is created, all the container statuses have `c.State.Waiting != nil`
1. A container manages to start up faster than the rest (e.g. `pod.Status.ContainerStatuses[2].State.Waiting == nil` while the rest are still non-nil) and a `watch.Modified` event gets sent out
1. `containers[pod.Spec.Containers[2].Name]` is set and sent to the `readyContainers` chan
1. the client starts streaming for container[2], but that container is blocked until container[0] and [1] completes
1. the final printed logs will be from containers 2 -> 0 -> 1 -> 3 -> 4 -> ...

Reply:
ya this might be possible, I haven't been able to produce it myself, but in theory this could happen. Maybe there is another check I can do